TestingHomework-Discounts/Accessors/CheckoutAccessor.cs

```diff
@@ -0,0 +1,201 @@
+                if (dbPromo != null)
+                {
+                    DateTime now = DateTime.UtcNow;
+                    if (now < dbPromo.StartDate)
+                    {
+                        cart.PromoErrors.Add(new PromoError()
+                        {
+                            ErrorCode = PromoErrorType.NotStarted,
+                            Message = $"Promo starts on {dbPromo.StartDate}"
+                        });
+                    }
+                    if (dbPromo.EndDate < now)
+                    {
+                        cart.PromoErrors.Add(new PromoError()
+                        {
+                            ErrorCode = PromoErrorType.Expired,
+                            Message = $"Promo expired"
+                        });
+                    }
+                    if (dbPromo.RedemptionCount >= dbPromo.MaxRedemptionCount)
+                    {
+                        cart.PromoErrors.Add(new PromoError()
+                        {
+                            ErrorCode = PromoErrorType.MaxRedemptions,
+                            Message = $"Promo expired"
+                        });
+                    }
+
+                    if (cart.Products.Select(_product => _product.Id).Contains(dbPromo.Id) == false)
+                    {
+                        cart.PromoErrors.Add(new PromoError()
+                        {
+                            ErrorCode = PromoErrorType.NoRelatedProduct,
+                            Message = "Promo does not apply to any items in the cart"
+                        });
+                    }
+
+                    dbPromo.RedemptionCount += 1;
+                    db.SaveChanges();
+
+                    // apply promo
+                    cart.PromoCodes.Add(dbPromo);
+                }
+                else
+                {
+                    cart.PromoErrors.Add(new PromoError()
+                    {
+                        ErrorCode = PromoErrorType.InvalidPromo,
+                        Message = $"Promo {code} doesn't exist"
+                    });
+                }
+
+                return cart;
+            }
+
+        }
+
+        public Cart GetUserCart(Guid userId)
+        {
+            using (PromoRepository db = new PromoRepository())
+            {
+                Cart userCart = db.Carts
+                    .Include(_cart => _cart.User)
+                    .FirstOrDefault(_cart => _cart.User.Id == userId);
+
+                userCart.Products = db.CartProducts.Include(cp => cp.Product).Where(cp => cp.CartId == userCart.Id).Select(cp => cp.Product).ToList();
+                userCart.PromoCodes = db.CartPromos.Include(cp => cp.PromoCode).Where(cp => cp.CartId == userCart.Id).Select(cp => cp.PromoCode).ToList();
+
+                return userCart;
+            }
+        }
+            
+        public Cart SaveCart(Cart cart)
+        {
+            using (PromoRepository db = new PromoRepository())
```

Comment:
There should definitely be some refactoring in this method. This method is mixing responsibilities such as
 - checking for product changes on a cart
 - saving products
 - managing related promotions
 - saving a cart